### ISSUE TITLE
Bug: Permanently deleting a 'simple' product while it is a child of 'grouped' product does not update the grouped product's '_children' postmeta

### ISSUE THREAD

**Issue by hjoelr** [woocommerce/woocommerce#17132]
I ran into an issue the other day where permanently deleting a `simple` product that also happened to be included in a `grouped` product's 'Grouped products' section would not delete the child product from the `_children` array of the `grouped` product.

I thought maybe it could have been some plugin interfering with this, so I disabled every plugin except WooCommerce and permanently deleted a product that fit the parameters above and the `_children` postmeta was still not updated correctly. That leads me to believe that this is a bug.

I did search for other issues and came across [this one](https://github.com/woocommerce/woocommerce/issues/14217) that seems to be related but I think is different. That one is more of a feature request than a bug report.

## Prerequisites

- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate
- [x] The issue still exists against the latest `master` branch of WooCommerce on Github (this is **not** the same version as on WordPress.org!)
- [x] I have attempted to find the simplest possible steps to reproduce the issue
- [ ] I have included a failing test as a pull request (Optional)

## Steps to reproduce the issue

1. Create a **Simple** product with just a title of 'Test Simple'
2. Create another **Grouped** product with just a title of 'Test Grouped' and add the 'Test Simple' product to the **Linked Products** > **Grouped products** section.
3. Confirm that the 'Test Grouped' product in the DB has a `_children` postmeta that looks like this:

```
a:1:{i:0;i:39498;}
```

4. Move the 'Test Simple' product to the Trash.
5. Go to the product trash and permanently delete the 'Test Simple' product.
6. Look at the `_children` postmeta on the 'Test Grouped' product and notice that the 'Test Simple' product's ID has not been removed.

## Expected/actual behavior

I expected the `_children` postmeta of the 'Test Grouped' product to no longer have the ID of the 'Test Simple' product.

## Isolating the problem

- [x] This bug happens with only WooCommerce plugin active
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/)
- [x] I can reproduce this bug consistently using the steps above

## WordPress Environment

<details>

```
### WordPress Environment ###

Home URL: http://localdev.dev/store
Site URL: http://localdev.dev/store
WC Version: 3.2.0
Log Directory Writable: ✔
WP Version: 4.8.2
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: –
Language: en_US

### Server Environment ###

Server Info: Apache/2.4.10 (Debian)
PHP Version: 7.1.5
PHP Post Max Size: 8 MB
PHP Time Limit: 30
PHP Max Input Vars: 1000
cURL Version: 7.38.0
OpenSSL/1.0.1t

SUHOSIN Installed: –
MySQL Version: 5.7.17
Max Upload Size: 2 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ❌ Your server does not have the SoapClient class enabled - some gateway plugins which use SOAP may not work as expected.
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.1.2
WC Database Prefix: ce0p3_
Total Database Size: 116.32MB
Database Data Size: 77.95MB
Database Index Size: 38.37MB
ce0p3_woocommerce_sessions: Data: 0.05MB + Index: 0.02MB
ce0p3_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB
ce0p3_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB
ce0p3_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.05MB
ce0p3_woocommerce_order_items: Data: 0.06MB + Index: 0.02MB
ce0p3_woocommerce_order_itemmeta: Data: 0.22MB + Index: 0.27MB
ce0p3_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB
ce0p3_woocommerce_tax_rate_locations: Data: 0.08MB + Index: 0.14MB
ce0p3_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB
ce0p3_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.05MB
ce0p3_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB
ce0p3_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB
ce0p3_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB
ce0p3_woocommerce_log: Data: 0.02MB + Index: 0.02MB
ce0p3_akeeba_common: Data: 0.02MB + Index: 0.00MB
ce0p3_ak_params: Data: 0.02MB + Index: 0.00MB
ce0p3_ak_profiles: Data: 0.02MB + Index: 0.00MB
ce0p3_ak_stats: Data: 0.11MB + Index: 0.03MB
ce0p3_ak_storage: Data: 11.52MB + Index: 0.00MB
ce0p3_ak_users: Data: 0.02MB + Index: 0.00MB
ce0p3_commentmeta: Data: 0.02MB + Index: 0.03MB
ce0p3_comments: Data: 0.09MB + Index: 0.09MB
ce0p3_links: Data: 0.02MB + Index: 0.02MB
ce0p3_options: Data: 3.19MB + Index: 0.19MB
ce0p3_pmxe_exports: Data: 0.33MB + Index: 0.00MB
ce0p3_pmxe_posts: Data: 1.52MB + Index: 0.00MB
ce0p3_pmxe_templates: Data: 0.02MB + Index: 0.00MB
ce0p3_pmxi_files: Data: 0.02MB + Index: 0.00MB
ce0p3_pmxi_history: Data: 0.02MB + Index: 0.00MB
ce0p3_pmxi_imports: Data: 1.52MB + Index: 0.00MB
ce0p3_pmxi_posts: Data: 1.52MB + Index: 0.00MB
ce0p3_pmxi_templates: Data: 0.09MB + Index: 0.00MB
ce0p3_postmeta: Data: 22.56MB + Index: 19.06MB
ce0p3_posts: Data: 27.56MB + Index: 7.58MB
ce0p3_relevanssi: Data: 2.52MB + Index: 6.06MB
ce0p3_relevanssi_log: Data: 0.06MB + Index: 0.00MB
ce0p3_relevanssi_stopwords: Data: 0.02MB + Index: 0.00MB
ce0p3_termmeta: Data: 0.09MB + Index: 0.16MB
ce0p3_terms: Data: 0.09MB + Index: 0.14MB
ce0p3_term_relationships: Data: 3.52MB + Index: 3.52MB
ce0p3_term_taxonomy: Data: 0.09MB + Index: 0.14MB
ce0p3_usermeta: Data: 0.11MB + Index: 0.11MB
ce0p3_users: Data: 0.02MB + Index: 0.05MB
ce0p3_vtprd_lifetime_limits_purchaser: Data: 0.02MB + Index: 0.02MB
ce0p3_vtprd_lifetime_limits_purchaser_logid_rule: Data: 0.02MB + Index: 0.02MB
ce0p3_vtprd_lifetime_limits_purchaser_rule: Data: 0.02MB + Index: 0.02MB
ce0p3_vtprd_purchase_log: Data: 0.23MB + Index: 0.02MB
ce0p3_vtprd_purchase_log_product: Data: 0.02MB + Index: 0.02MB
ce0p3_vtprd_purchase_log_product_rule: Data: 0.02MB + Index: 0.02MB
ce0p3_woocommerce_bundled_itemmeta: Data: 0.23MB + Index: 0.31MB
ce0p3_woocommerce_bundled_items: Data: 0.02MB + Index: 0.03MB
ce0p3_wpfront_ure_login_redirect: Data: 0.02MB + Index: 0.00MB
ce0p3_wpfront_ure_options: Data: 0.03MB + Index: 0.00MB

### Post Type Counts ###

attachment: 2187
banner: 1
banner_product: 4
cue_playlist: 52
customize_changeset: 6
nav_menu_item: 45
page: 17
post: 3
product: 2744
product_variation: 2
revision: 261
shop_coupon: 11
shop_order: 149
vtprd-rule: 8
wishlist: 9

### Security ###

Secure connection (HTTPS): ❌Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ❌Error messages should not be shown to visitors.

### Active Plugins (1) ###

WooCommerce: by Automattic – 3.2.0-rc.2

### Settings ###

API Enabled: ✔
Force SSL: ✔
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: bundle (bundle)
external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)


### WC Pages ###

Shop base: #181 - /shop/
Cart: #191 - /cart/
Checkout: #701 - /checkout/
My account: #711 - /my-account/
Terms and conditions: #8543 - /terms-conditions/

### Theme ###

Name: Custom Theme
Version: 1.1.9.0
Child Theme: ✔
Parent Theme Name: Storefront
Parent Theme Version: 2.2.5
Parent Theme Author URL: https://woocommerce.com/
WooCommerce Support: ✔

### Templates ###

Overrides: custom-theme/woocommerce/archive-product.php
custom-theme/woocommerce/cart/cart.php
custom-theme/woocommerce/checkout/form-coupon.php
custom-theme/woocommerce/checkout/payment-method.php
custom-theme/woocommerce/checkout/review-order.php
custom-theme/woocommerce/emails/email-customer-details.php
custom-theme/woocommerce/global/quantity-input.php version 2.5.0 is out of date. The core version is 3.2.0
custom-theme/woocommerce/loop/add-to-cart.php
custom-theme/woocommerce/loop/orderby.php
custom-theme/woocommerce/loop/result-count.php
custom-theme/woocommerce/single-product/add-to-cart/grouped.php
custom-theme/woocommerce/single-product/title.php

Outdated Templates: ❌Learn how to update
```

</details>

**Comment by claudiulodro** [comment]
This is limited by the same problem #14217 has:
> There is no way to determine the parent, because there can be multiple parents.

We need to implement a custom table for parent/child relationships to accurately and efficiently get the parents in order to update them when a child is deleted, otherwise we're just trying to query for partial string matches in the serialized data to figure out what a child's parents are.

Currently, deleted children won't show up on any admin/customer pages and re-saving the grouped product from its admin page will update the children and remove any deleted ones.

I'm closing this issue, but it will be handled when we implement relationship tables. Good catch.

**Comment by hjoelr** [comment]
@claudiulodro Yeah, the custom table is probably the right approach.

The reason I found this issue was because of two different plugins I was using (one we created internally and one we purchased) that weren't expecting to get a `false` returned from `wc_get_product( $child_id )` as they are enumerating through the `$grouped_product->get_children()` results. Both of them assumed that `get_children()` would return real product IDs. So, even though deleted children may not be showing up on any admin or customer pages, this bug still has some implications.

I saw in #14217 that creating the relationship table isn't planned until WC 4.0. I'm wondering if there's any way it could be implemented sooner.
